### ISSUE TITLE
Scripts for Ego-Exo Correspondence

### ISSUE THREAD

**Issue by RZkiller** [zamling/PSALM#18]
Hi, thank you for the great job!
 I noticed that PSALM has capability to handle Ego-exo correspondence in Ego-Exo4d and the paper shows results for that, but the repository doesn't seem to provide scripts for the task. I'm wondering if you plan to release scripts for that, and if not, can you provide some hints about how to evaluate or train on the Ego-exo correspondence using PSALM?
 Thanks in advance!

**Comment by zamling** [comment]
@RZkiller 
Hi, We evaluated Ego-Exo4d (also video segmentation) by interactive segmentation schema. Specifically, we use the mask in ego (exo) view to generate visual prompt embedding and predict exo (ego) view mask. You can change the input of interactive segmentation to adapt Ego-Exo4d.

**Comment by zamling** [comment]
There is a better way! You can directly follow the video object segmentation [script](https://github.com/zamling/PSALM/blob/main/psalm/eval/eval_davis.py).

You need to adapt dataloader to Ego-Exo4d task, like
https://github.com/zamling/PSALM/blob/1f7eece2f3c3bb8e0d07e4c84835ec278315fceb/psalm/eval/eval_davis.py#L286

The `vp_annotations` and `vp_image` are the ego (exo) mask prompt and image, while `file_name` and `annotations` are exo (ego) GT and image

**Comment by RZkiller** [comment]
@zamling 
Thank you for your reply! I have a few more questions about the implementation of Ego-Exo Correspondence:
1. Your reply mentions the need to adapt dataloader to Ego-Exo4d task, but I noticed that Davis_Dataset and its inherited parent class COCO_interactive_dataset didn't particularly need to be changed. Does this mean we just need a json file like trainval_val_psalm.json? If not, can you tell me the exact places that need to be changed?
2. Do I need to set up new coco_class_name and coco_class_ids for the Ego-Exo4d dataset? Or just follow the coco_class_name and coco_class_ids provided by the parent class COCO_interactive_dataset?

**Comment by zamling** [comment]
@RZkiller 
For dataset load, it seems that you only need to rewrite the IO to define the vp_image and vp_annotation. But we only write the inference forward for davis, so that you also need to rewrite the training codes for Ego-Exo. The main modification may be `prepare_inputs_labels_for_multimodal` function like
https://github.com/zamling/PSALM/blob/1f7eece2f3c3bb8e0d07e4c84835ec278315fceb/psalm/model/language_model/llava_phi.py#L1639

And others follow interactive coco training.

**Comment by RZkiller** [comment]
Following your method, I have implemented ego-exo correspond evaluation(ego2exo setting), but the result is totally not good. 
![2220](https://github.com/user-attachments/assets/42185d3e-f55a-480f-b70c-4c0a7c83b6a8)
![330](https://github.com/user-attachments/assets/d45f8534-93ae-4a91-a58b-cf3b1347da43)
As shown in the pictures, some frames do not segment the object at all, and some frames segment not the object of interest (in this case is basketball)

So I still have some questions about the implementation of ego-exo evaluation:
1. I noticed that the size of vp_image in the original davis inference script is the same as the size of the subsequent predicted frame. But in the ego-exo4d dataset, exo ! =ego size, i.e. the size of vp_image is not the same as the size of the subsequently predicted exo frames. I'd like to know if this has any effect on the results, or can you tell me how you dealt with this issue
2. If I choose the first frame of ego as vp_image, should the predicted exo start from the first frame or the second frame?
[https://github.com/zamling/PSALM/blob/main/datasets/build_DAVIS.py#L58](url)
Because I found that for the same predicted frame, if exo start from the first frame then the ball is not segmented. If exo start from the second frame then it can be segmented. So how to set the start frame of exo?
![330](https://github.com/user-attachments/assets/f8f1aede-1de5-4bf8-a6fd-fd4c67d6926b)
![Copy of 330](https://github.com/user-attachments/assets/3263fb80-3a2e-46fe-91ce-4f62e07e7a3e)

**Comment by zamling** [comment]
Hi @RZkiller 
For your questions:
1. The different size of image and vp_image does not affect the results. PSALM solve interactive segmentation (also video object segmentation and Ego-Exo4D) by mask pooling vp_image features, rather than absolute position info. So that you can find that when you want to segment a ball, PSALM will segment the light in the ceiling in the second figure you showed.
2. It should be the first frame.

We only download a lower resolution images (may be 448P) and rebuild the Ego-Exo4D data. But the rebuilt data is on another machine and I can not use it when I want to refactor this code. Also this is the reason why I can not release the evaluation pipline for Ego-Exo4D... 

Ego-Exo4D is only a zero-shot test in our method and also we only evaluate it on val set, as shown in Tab. 18. You can see that the performance is not good enough. So may be you need to ft or make extra designs for this task, rather than use interactive segmentation pipline.

In addition, I can provide you the __getitem__ method for Ego-Exo4D, although I foget its data format...
```
    def __getitem__(self, idx):
        ego_query = False
        data = self.data[idx]
        image_root = os.path.join(self.data_args.image_folder,data['root_dir'],'frame_aligned_videos/downscaled/448')
        image_frame = str(data['take_frame']) + '.jpg'
        image_path = None
        vp_image_path = None
        vp_mask_list = []
        gt_mask_list = []
        object_pool = []
        object_list = data['object_list']
        for object in object_list:
            for key, value in data[object].items():
                cur_object_info = (os.path.join(image_root,key,image_frame),value)
                object_pool.append(cur_object_info)

        if ego_query:
            for obj in object_pool:
                name, mask = obj
                if 'ari' not in name:
                    if image_path is None:
                        image_path = name
                    else:
                        assert image_path == name, f'exsit name is {image_path}, while also a name {name}'
                    gt_mask_list.append(mask)
                else:
                    if vp_image_path is None:
                        vp_image_path = name
                    else:
                        assert vp_image_path == name, f'exsit vp name is {image_path}, while also a name {name}'
                    vp_mask_list.append(mask)
        else:
            for obj in object_pool:
                name, mask = obj
                if 'ari' in name:
                    if image_path is None:
                        image_path = name
                    else:
                        assert image_path == name, f'exsit name is {image_path}, while also a name {name}'
                    gt_mask_list.append(mask)

                else:
                    if vp_image_path is None:
                        vp_image_path = name
                    else:
                        assert vp_image_path == name, f'exsit vp name is {image_path}, while also a name {name}'
                    vp_mask_list.append(mask)

        # resize mask
        if not os.path.exists(image_path) or not os.path.exists(vp_image_path):
            print(f'cannot find {image_path}')
            return {}
        # image = cv2.imread(image_path)
        # h,w = image.shape[:2]
        # vp_image = cv2.imread(vp_image_path)
        # vp_h,vp_w = vp_image.shape[:2]
        gt_mask_h,gt_mask_w = decode_mask(gt_mask_list[0]).shape[:2]
        vp_mask_h,vp_mask_w = decode_mask(vp_mask_list[0]).shape[:2]
        img = utils.read_image(image_path, format='RGB')
        vp_img = utils.read_image(vp_image_path, format='RGB')
        img = cv2.resize(img,(gt_mask_w,gt_mask_h))
        vp_img = cv2.resize(vp_img,(vp_mask_w,vp_mask_h))
        # gt_mask_list = [decode_mask(mask_) for mask_ in gt_mask_list]
        # vp_mask_list = [decode_mask(mask_) for mask_ in vp_mask_list]
        data_dict = {}
        data_dict['file_path'] = image_path
        data_dict['gt_mask_list'] = copy.deepcopy(gt_mask_list)
        data_dict['vp_file_path'] = vp_image_path
        data_dict['file_name'] = img
        data_dict['vp_image'] = vp_img
        data_dict['height'] = gt_mask_h
        data_dict['width'] = gt_mask_w
        data_dict['image_id'] = idx
        data_dict['annotations'] = []
        data_dict['vp_annotations'] = []
        for mask in gt_mask_list:
            anno = {}
            anno['bbox_mode'] = BoxMode.XYXY_ABS
            anno['bbox'] = [0, 0, 0, 0]
            anno['image_id'] = idx
            anno['category_id'] = 1
            anno['segmentation'] = mask
            data_dict['annotations'].append(anno)
        for mask in vp_mask_list:
            anno = {}
            anno['bbox_mode'] = BoxMode.XYXY_ABS
            anno['bbox'] = [0, 0, 0, 0]
            anno['image_id'] = idx
            anno['category_id'] = 1
            anno['segmentation'] = mask
            data_dict['vp_annotations'].append(anno)
        try:
            if isinstance(self.data_args.image_processor,dict):
                processor = self.data_args.image_processor['instance']
            else:
                processor = self.data_args.image_processor
            region_mask_type = getattr(self.data_args,'region_mask_type',None)
            if region_mask_type is not None:
                region_mask_type = region_mask_type.split('||')
            data_dict = processor.preprocess(data_dict,region_mask_type=region_mask_type,mask_format='bitmask')
        except:
            print('load data wrong')
            return {}

        num_target = len(data_dict['instances'])
        prefix_inst = 'This is an image <image>, Please segment by given regions'
        # prompt_inst = 'Iteractive segmentation: using provided region as a reference'
        regions_inst = ' <region>,' * (num_target - 1) + ' <region>.'
        sources_value = f'\nThis is all regions: {regions_inst}\n'

        if self.data_args.seg_last:
            sources = [
                [{'from': 'human', 'value': prefix_inst + sources_value},
                 {'from': 'gpt', 'value': '\n[SEG]<seg>'}]]
        else:
            sources = [
                [{'from': 'human', 'value': prefix_inst + sources_value},
                 {'from': 'gpt', 'value': '\n[SEG]'}]]
        # sources = self.preprocess_multimodal(copy.deepcopy(sources))

        text_dict = self.preprocess_llama2(sources, self.tokenizer)
        input_ids = text_dict['input_ids'][0]
        labels = text_dict['labels'][0]
        data_dict['input_ids'] = input_ids
        data_dict['labels'] = labels
        data_dict['dataset_type'] = 'region_coco'

        return data_dict
```
